### PR TITLE
Faster video scan approach

### PR DESCRIPTION
**Pull request by jeannich**
## Description
This PR is an attempt to improve video scan by solving 2 issues:
1) progress bar report:
All videos are listed at the beginning of the scan, then videos are scanned sequentially which allows to report a relevant progress bar status that is not anymore constantly reverted to zero percent at each new folder scanned.

2) scan performance:
The function "DoScan" is not recursive anymore and does not repeat tasks anymore (it was previously, in some cases, re-listing over and over video medias.). Then it detects faster if a media is already in the DB or if it must be scanned.

NOTE: I'm a new contributor so please let me know if I missed some things.

## Motivation and context
I was not happy seeing kodi library update taking forever on my setup. I looked for solutions and found that other users were experimenting the same issues.

## How has this been tested?
I could not find test framework in kodi so I created a fake setup environment simulating the presence of 460 video items (movies and TV Shows) 
I compiled kodi from Omega branch and from this PR and compared the scan performance of both builds on the same hardware.

Note: After the first initial scan I got around 70% of videos recognized so next scans are constantly trying to fetch medias infos of 30% of the library.
At this stage, here are the results of "Update library":
Omega build: 112s
This PR build:  34s

Performance gain should be bigger for users having most of their medias already recognized as this algorithm is very fast at skipping medias already in DB.

## What is the effect on users?
Faster video scan performance.

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by neo1973** [comment]
Hi @jeannich, thanks for giving this a try. For starters, can you please apply the formatting patch from the Jenkins comment?

File: xbmc/video/VideoInfoScanner.cpp [hunk 1]
```diff
@@ -121,36 +121,45 @@ namespace VIDEO
...
-      bool bCancelled = false;
-      while (!bCancelled && !m_pathsToScan.empty())
+
+      if (m_handle)
+        m_handle->SetTitle("Listing media files ...");
+
+      // List all media files of all 'm_pathsToScan'
+      //   INFO: m_pathsToScan may contain path with different library content, so we store their properties into a dedicated struct.
+      std::vector<SourcePathContent> _pathsContent;
+      CFileItemList remoteDirectoryItems[m_pathsToScan.size()];
```
Comment (neo1973) [inline xbmc/video/VideoInfoScanner.cpp:131]: Variable sized arrays are not part of the C++ standard, please replace this with a `std::vector`.

File: xbmc/video/VideoInfoScanner.cpp [hunk 2]
```diff
@@ -121,36 +121,45 @@ namespace VIDEO
       // result in unexpected behaviour.
       m_bCanInterrupt = false;
 
-      bool bCancelled = false;
-      while (!bCancelled && !m_pathsToScan.empty())
+
+      if (m_handle)
+        m_handle->SetTitle("Listing media files ...");
```
Comment (ksooo) [inline xbmc/video/VideoInfoScanner.cpp:126]: This needs to be a localized string. Look for `g_localizeStrings.Get` usage to get an idea how this works.

Hint: you deleted a line ` m_handle->SetText(g_localizeStrings.Get(20415));` in this PR. Those numbers come from af ile named string.po you can easily find in our code base I guess.

File: xbmc/video/VideoInfoScanner.cpp [hunk 3]
```diff
@@ -121,36 +121,45 @@ namespace VIDEO
...
-         * A copy of the directory path is used because the path supplied is
-         * immediately removed from the m_pathsToScan set in DoScan(). If the
-         * reference points to the entry in the set a null reference error
-         * occurs.
-         */
-        std::string directory = *m_pathsToScan.begin();
-        if (m_bStop)
+        // Load root path properties from scraper settings:
+        bool foundDirectly = false;
+        bool bSkip = false;
```
Comment (ksooo) [inline xbmc/video/VideoInfoScanner.cpp:138]: coding guidelines: please remove the type prefix from the vriables' name. (https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#88-variables)

File: xbmc/video/VideoInfoScanner.cpp [hunk 4]
```diff
@@ -121,36 +121,45 @@ namespace VIDEO
...
+        SScanSettings rootPathScraperScanSettings;
+        ScraperPtr info = m_database.GetScraperForPath(rootPathToScan, rootPathScraperScanSettings, foundDirectly);
+        CONTENT_TYPE contentType_AsDefinedInScraper = info ? info->Content() : CONTENT_NONE;
```
Comment (ksooo) [inline xbmc/video/VideoInfoScanner.cpp:141]: As a general rule, if you assign a value only once during lifetime of a variable, make it `const` to make this fact clear to the reader of the code.

File: xbmc/video/VideoInfoScanner.cpp [hunk 5]
```diff
@@ -121,36 +121,45 @@ namespace VIDEO
...
+
+        // List and store medias:
+        if( isPathAllowedForScan(rootPathToScan, rootPathScraperScanSettings, contentType_AsDefinedInScraper))
```
Comment (ksooo) [inline xbmc/video/VideoInfoScanner.cpp:144]: coding guidelines: Function names need to start with an uppercase letter. (https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#87-methods)

File: xbmc/video/VideoInfoScanner.cpp [hunk 6]
```diff
@@ -121,36 +121,45 @@ namespace VIDEO
...
         {
-          bCancelled = true;
+          loadRemoteItems(rootPathToScan, remoteDirectoryItems[nbRemoteDirectoryItems++], rootPathScraperScanSettings, contentType_AsDefinedInScraper);
```
Comment (ksooo) [inline xbmc/video/VideoInfoScanner.cpp:146]: coding guidelines: Function names need to start with an uppercase letter. (https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#87-methods)

File: xbmc/video/VideoInfoScanner.cpp [hunk 7]
```diff
@@ -121,36 +121,45 @@ namespace VIDEO
...
+          _pathsContent.push_back({rootPathToScan, rootPathScraperScanSettings, contentType_AsDefinedInScraper, remoteDirectoryItems[nbRemoteDirectoryItems-1]});
+
+          //We use the 'path' of CFileItemList to store its root-path-to-scan
```
Comment (ksooo) [inline xbmc/video/VideoInfoScanner.cpp:149]: Please insert a blank after `//` for better readability

File: xbmc/video/VideoInfoScanner.cpp [hunk 8]
```diff
@@ -121,36 +121,45 @@ namespace VIDEO
...
-           * Note that this will skip clean (if m_bClean is enabled) if the directory really
-           * doesn't exist rather than a NAS being switched off.  A manual clean from settings
-           * will still pick up and remove it though.
-           */
-          CLog::Log(LOGWARNING, "{} directory '{}' does not exist - skipping scan{}.", __FUNCTION__,
-                    CURL::GetRedacted(directory), m_bClean ? " and clean" : "");
-          m_pathsToScan.erase(m_pathsToScan.begin());
+
+        // Support scan cancellation
+        if (m_bStop) {
```
Comment (ksooo) [inline xbmc/video/VideoInfoScanner.cpp:154]: coding guidelines: put `{` on next line. (https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#31-braces) - here and everywhere else in your PR.

File: xbmc/video/VideoInfoScanner.cpp [hunk 9]
```diff
@@ -237,201 +246,159 @@ namespace VIDEO
...
-    ScraperPtr info = m_database.GetScraperForPath(strDirectory, settings, foundDirectly);
-    CONTENT_TYPE content = info ? info->Content() : CONTENT_NONE;
+          std::string remoteItemPath = remoteItem->GetPath();
+          loadRemoteItems(remoteItemPath, remoteDirectoryItems, rootPathScraperScanSettings, contentType_AsDefinedInScraper);
+        }
+      } else { // For other media types, it's the reverse: we only store files
+        if (remoteItem->m_bIsFolder)
+        {
+          if(rootPathScraperScanSettings.recurse > 0 ) {
+            if( isPathAllowedForScan(strDirectory, rootPathScraperScanSettings, contentType_AsDefinedInScraper) ) {
```
Comment (ksooo) [inline xbmc/video/VideoInfoScanner.cpp:280]: coding guidelines: blank after`if`, no blank after `(` please (maybe this is just a typo)

File: xbmc/video/VideoInfoScanner.cpp [hunk 10]
```diff
@@ -237,201 +246,159 @@ namespace VIDEO
...
+              std::string remoteItemPath = remoteItem->GetPath();
+              loadRemoteItems(remoteItemPath, remoteDirectoryItems, rootPathScraperScanSettings, contentType_AsDefinedInScraper);
```
Comment (ksooo) [inline xbmc/video/VideoInfoScanner.cpp:282]: you could remove the variable as you use it only once.

File: xbmc/video/VideoInfoScanner.cpp [hunk 11]
```diff
@@ -237,201 +246,159 @@ namespace VIDEO
...
-        item->m_bIsFolder = true;
-        items.Add(item);
-        items.SetPath(URIUtils::GetParentPath(item->GetPath()));
-      }
-    }
-    bool foundSomething = false;
-    if (!bSkip)
+  bool CVideoInfoScanner::DoScan(const std::string& strDirectory){
+    return false;
+  }
```
Comment (ksooo) [inline xbmc/video/VideoInfoScanner.cpp:329]: Can you explain this change please?

File: xbmc/video/VideoInfoScanner.cpp [hunk 12]
```diff
@@ -237,201 +246,159 @@ namespace VIDEO
...
+  bool CVideoInfoScanner::DoScan(const std::vector<SourcePathContent>& _pathsContent)
```
Comment (ksooo) [inline xbmc/video/VideoInfoScanner.cpp:330]: coding guidelines: parameter names must start with a lower case letter (https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#87-methods)

File: xbmc/video/VideoInfoScanner.cpp [hunk 13]
```diff
@@ -237,201 +246,159 @@ namespace VIDEO
...
-        if (m_bClean)
-          m_pathsToClean.insert(m_database.GetPathId(strDirectory));
-        CLog::Log(LOGDEBUG, "VideoInfoScanner: No (new) information was found in dir {}",
-                  CURL::GetRedacted(strDirectory));
-      }
-    }
-    else if (!StringUtils::EqualsNoCase(hash, dbHash) && (content == CONTENT_MOVIES || content == CONTENT_MUSICVIDEOS))
-    { // update the hash either way - we may have changed the hash to a fast version
-      m_database.SetPathHash(strDirectory, hash);
+      m_handle->SetTitle(g_localizeStrings.Get(20415));  //"Scanning for new content"
```
Comment (ksooo) [inline xbmc/video/VideoInfoScanner.cpp:334]: Insert a blank after `//` please

File: xbmc/video/VideoInfoScanner.cpp [hunk 14]
```diff
@@ -237,201 +246,159 @@ namespace VIDEO
...
 
-      // add video extras to library
-      if (foundSomething && !m_ignoreVideoExtras && pItem->IsVideoExtras())
+      for (int i = 0; i < remoteDirectoryItems.Size(); ++i)
       {
-        if (AddVideoExtras(items, content, pItem->GetPath()))
-        {
-          CLog::Log(LOGDEBUG, "VideoInfoScanner: Finished adding video extras from dir {}",
-                    CURL::GetRedacted(pItem->GetPath()));
+        CFileItemPtr remoteItem = remoteDirectoryItems[i];
```
Comment (ksooo) [inline xbmc/video/VideoInfoScanner.cpp:350]: Please use the real type `std::shared_ptr<CFileItem>`. We want to get rid of the `CFileItemPtr` typedef.

File: xbmc/video/VideoInfoScanner.cpp [hunk 15]
```diff
@@ -237,201 +246,159 @@ namespace VIDEO
...
+        if (m_bStop)
+          return !m_bStop;
 
-      // if we have a directory item (non-playlist) we then recurse into that folder
-      // do not recurse for tv shows - we have already looked recursively for episodes
-      if (pItem->m_bIsFolder && !pItem->IsParentFolder() && !pItem->IsPlayList() && settings.recurse > 0 && content != CONTENT_TVSHOWS)
-      {
-        if (!DoScan(pItem->GetPath()))
+
+        if( (pathContent.contentType_AsDefinedInScraper == CONTENT_TVSHOWS && remoteItem->m_bIsFolder) //For TV shows, we only scan folders
```
Comment (ksooo) [inline xbmc/video/VideoInfoScanner.cpp:365]: Insert blank after `//`

File: xbmc/video/VideoInfoScanner.cpp [hunk 16]
```diff
@@ -237,201 +246,159 @@ namespace VIDEO
...
+            case CONTENT_MOVIES:
+              isVideoAlreadyInDB = m_database.HasMovieInfo(remoteItemPath);
+              break;
+            case CONTENT_MUSICVIDEOS:
+              isVideoAlreadyInDB = m_database.HasMusicVideoInfo(remoteItemPath);
+              break;
+            case CONTENT_TVSHOWS:
+              isVideoAlreadyInDB = m_database.HasTvShowInfo(remoteItemPath);
+              break;
+          }
```
Comment (ksooo) [inline xbmc/video/VideoInfoScanner.cpp:382]: I would insert a `default` label to indicate that all intended value are handled here. as far as I know there are more types than only those handled here. Compiler should even generate a warning that you do not handle all possible values here?

File: xbmc/video/VideoInfoScanner.cpp [hunk 17]
```diff
@@ -471,23 +438,20 @@ namespace VIDEO
         continue;
 
       // Discard all exclude files defined by regExExclude
-      if (CUtil::ExcludeFileOrFolder(pItem->GetPath(), (content == CONTENT_TVSHOWS) ? CServiceBroker::GetSettingsComponent()->GetAdvancedSettings()->m_tvshowExcludeFromScanRegExps
+      //         TODO: probably not necessary anymore as replaced by 'isPathAllowedForScan' function
```
Comment (ksooo) [inline xbmc/video/VideoInfoScanner.cpp:441]: Use `//! @todo` for todos please

File: xbmc/video/VideoInfoScanner.cpp [hunk 18]
```diff
@@ -471,23 +438,20 @@ namespace VIDEO
...
-      }
 
       // clear our scraper cache
       info2->ClearCache();
 
       INFO_RET ret = INFO_CANCELLED;
       if (info2->Content() == CONTENT_TVSHOWS)
         ret = RetrieveInfoForTvShow(pItem.get(), bDirNames, info2, useLocal, pURL, fetchEpisodes, pDlgProgress);
-      else if (info2->Content() == CONTENT_MOVIES)
+      else if (info2->Content() == CONTENT_MOVIES) // if (pItem->m_bIsFolder
```
Comment (ksooo) [inline xbmc/video/VideoInfoScanner.cpp:454]: remove the comment

**Review by ksooo — COMMENTED** [review]
Thank you very much for your contribution. I rushed over the code changes and would ask you to apply the changes I requested. Mostly coding guidelines compliance things.

I encourage you to read our coding guidelines: https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md

**Comment by jeannich** [comment]
Thanks a lot @ksooo & @neo1973 for your extensive review.
I'm not a C++ expert so bare with me for all those fixes to be applied.

I'll apply your recommendations and will update the PR in a few hours/days

**Comment by CrystalP** [comment]
Thanks for the contribution, that'd be a massive speed improvement!

After a quick read (haven't ran the code yet), I'm concerned about scraper settings retrieval. They may be set by the user on any directory at any level and affect all subpaths (until another overriding setting is found lower). The code likely works for an already populated library and all paths are provided in m_pathsToScan, but what about a new library, or the addition of sub folders with different scraper settings to an existing library?

The current code is inefficient wrt scrapers lookup/creation, and a local cache of scanners already used in the scan session could certainly help.

**Comment by jeannich** [comment]
Hi @CrystalP
I --think-- I handled this problematic

First let me define some keywords to make sure we understand each other:
A "video_source" can only contain one unique "scraper_settings"
And a "video_source" can contain multiple "root_path" (which, therefore, are also attached to only one "scraper_settings")

My understanding is that ALL subpath of a "root_path" share the same "scraper_settings".
If a root_path content is defined to "movies", there is no way for it's subpath to define that it contains something else.
(The only thing a subpath can halter is to contain a .nomedia file.)

==> Is that a correct understanding ? (If not, well I'll need to rethink many things ^^' )

**Comment by jeannich** [comment]
"They may be set by the user on any directory at any level and affect all subpaths"
Can you explain how can be done by a user? (I did not find this feature through the UI)


"The current code is inefficient wrt scrapers lookup/creation, and a local cache of scanners already used in the scan session could certainly help."
Could you detail here, not sure to get the point.

_(overall, is it the correct location to have all these messages ? I mean, fine with me, just to know how contributors here usually do as I feel like we could have many exchanges for this PR)_

**Comment by jjd-uk** [comment]
> "They may be set by the user on any directory at any level and affect all subpaths" Can you explain how can be done by a user? (I did not find this feature through the UI)
> 
Via Files view select any folder within the source open Context menu and you'll see Change content.

image

image

This allows you to change either the scraper use or the scraper settings for that particular folder/subfolders. For example there maybe a folder for movies from a different region requiring a different scrapper or settings from the default for the video source.

**Comment by jeannich** [comment]
Thanks @jjd-uk,
OK so what you described is a "video_source" called "Movie...." where you set the content type to "Movies".
Now this "video_source" contains one or more "root_path" right ? Where your content is stored.
Well, those root_path (including their sub-path of course), within this "video_source", will all be scanned for Movies only.

image
In my screenshot I have one "video_source" called "Movies samples", which contains 3 "root_path" (/Users/xxx/Desktop, Library and Movies).
Well those 3 root_path, and all their subpath, will only be scanned for the content defined in "Movies samples".

So, I think it works

**Comment by jjd-uk** [comment]
> Thanks @jjd-uk, OK so what you described is a "video_source" called "Movie...." where you set the content type to "Movies". Now this "video_source" contains one or more "root_path" right ? Where your content is stored. Well, those root_path (including their sub-path of course), within this "video_source", will all be scanned for Movies only.
> 
That is only one use case. 

Taking a structure

--Videos
---- Movies
------ Movies 1
-------- Movie A
-------- Movie B
------ Movies 2
-------- Movie C 
-------- Movie D
---- TV

Some of the possibilities are below.

Multipath (as in your example)

--Videos
---- Movies
------ Movies 1 (Source 1 Name = Movies / Content = Movies)
-------- Movie A
-------- Movie B
------ Movies 2 (Source 1 Name = Movies / Content = Movies)
-------- Movie C 
-------- Movie D
---- TV (Source 2 Name = TV Content = TV)

Separate source for each movie folder

--Videos
---- Movies
------ Movies 1 (Source 1 Name = Movies 1 / Content = Movies)
-------- Movie A
-------- Movie B
------ Movies 2  (Source 2 Name = Movies 2  / Content = Movies)
-------- Movie C 
-------- Movie D
---- TV (Source 3 Name = TV Content = TV)

Common source set to None with Content only set on content folders only

--Videos
---- Movies  (Source 1 Label =  Movies / Content = None)
------ Movies 1 ( Use Change content to set Content = Movies)
-------- Movie A
-------- Movie B
------ Movies 2 ( Use Change content to set Content = Movies)
-------- Movie C 
-------- Movie D
---- TV (Source 2 Name = TV Content = TV)

or

--Videos (Source 1 Label = Videos / Content = None)
---- Movies 
------ Movies 1 ( Use Change content to set Content = Movies)
-------- Movie A
-------- Movie B
------ Movies 2 ( Use Change content to set Content = Movies)
-------- Movie C 
-------- Movie D
---- TV (Source 2 Name = TV Content = TV)

or

Movie folder specific scraper/settings

--Videos
---- Movies
------ Movies 1 (Source 1 Name = Movies 1 / Content = Movies)
-------- Movie A ( Use Change content to chnage scraper or settings for only this folder)
-------- Movie B
------ Movies 2  (Source 2 Name = Movies 2  / Content = Movies)
-------- Movie C 
-------- Movie D 
---- TV (Source 2 Name = TV Content = TV)

**Comment by jeannich** [comment]
FYI I'm currently sick so updates will take some time to come but I'll do it.
Stay tuned

File: xbmc/video/VideoInfoScanner.cpp (hunk 11 above)
Comment (jeannich) [reply xbmc/video/VideoInfoScanner.cpp]: Yes well here is a problem indeed.
DoScan is a pure virtual function inherited from class CInfoScanner.

The problem is that the new scan logic is not doable with a simple "strDirectory" parameter as entry.
So I did not find a way to keep any purpose for this function but I cannot remove it without changing a lot of things such as not being a sub class of CInfoScanner.

This "DoScan" function is never called from outside so keeping this "dead" version is not great, but should not do harm.

Let me know if you see a better way.

File: xbmc/video/VideoInfoScanner.cpp (hunk 16 above)
Comment (jeannich) [reply xbmc/video/VideoInfoScanner.cpp]: Thanks for the comment.
I thought about this but found no action I would do in such "default" case.

If it's really mandatory to have one I would write:
```
default:
  isVideoAlreadyInDB = false;
  break;
```

Which does not make sense as isVideoAlreadyInDB is by default set to false anyway.
Not sure what I should change here                      _(and I did not check compiler's output)_

I would prefer to leave it like that, but feel free to give more orientations.

**Comment by jeannich** [comment]
Thanks for waiting, PR is now updated and ready for another review please !
@ksooo @jjd-uk @CrystalP @neo1973

File: xbmc/video/VideoInfoScanner.cpp [hunk 19]
```diff
@@ -2211,26 +2222,28 @@ namespace VIDEO
...
-  std::string CVideoInfoScanner::GetFastHash(const std::string &directory,
-      const std::vector<std::string> &excludes) const
+  std::string CVideoInfoScanner::GetFastHash(
+      const std::string& directory, const std::vector<std::string>& regexExcludeRules) const
   {
     CDigest digest{CDigest::Type::MD5};
 
-    if (excludes.size())
-      digest.Update(StringUtils::Join(excludes, "|"));
+    if (regexExcludeRules.size())
```
Comment (CastagnaIT) [inline xbmc/video/VideoInfoScanner.cpp:2245]:
```suggestion
    if (!regexExcludeRules.empty())
```

File: xbmc/video/VideoInfoScanner.cpp [hunk 20]
```diff
@@ -2247,17 +2260,17 @@ namespace VIDEO
...
   {
     CFileItemList items;
     items.Add(std::make_shared<CFileItem>(directory, true));
     CUtil::GetRecursiveDirsListing(directory, items, DIR_FLAG_NO_FILE_DIRS | DIR_FLAG_NO_FILE_INFO);
 
     CDigest digest{CDigest::Type::MD5};
 
-    if (excludes.size())
-      digest.Update(StringUtils::Join(excludes, "|"));
+    if (regexExcludeRules.size())
```
Comment (CastagnaIT) [inline xbmc/video/VideoInfoScanner.cpp:2272]:
```suggestion
    if (!regexExcludeRules.empty())
```

File: addons/resource.language.en_gb/resources/strings.po [hunk 21]
```diff
@@ -24345,4 +24345,8 @@ msgctxt "#40436"
 msgid "50th Anniversary Edition"
 msgstr ""
 
+msgctxt "#40437"
```
Comment (CastagnaIT) [inline addons/resource.language.en_gb/resources/strings.po:24348]: missing help comments for translators
```
#. [...describe meaning/use...]
#: xbmc/video/VideoInfoScanner.cpp
```

**Comment by jeannich** [comment]
> I've made some formatting changes to meet the current code style. The diffs are available in the following links:
> 
>     * [PR25964-0001-9d8eeb0438-Faster-video-scan-approach.diff](https://jenkins.kodi.tv/job/BuildMulti-PR/32152/artifact/PR25964-0001-9d8eeb0438-Faster-video-scan-approach.diff)
> 
>     * [PR25964-0003-f7330d51c0-Add-copy-move-constructors-to-be-h.diff](https://jenkins.kodi.tv/job/BuildMulti-PR/32152/artifact/PR25964-0003-f7330d51c0-Add-copy-move-constructors-to-be-h.diff)
> 
>     * [PR25964-0004-5934977d7f-Fixes-and-improvements.diff](https://jenkins.kodi.tv/job/BuildMulti-PR/32152/artifact/PR25964-0004-5934977d7f-Fixes-and-improvements.diff)
> 
> 
> For more information please see our [current code style guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md).

The diff suggested applies to an old version of the code so I cannot apply them.
Any suggestion how I can fix that ?

**Comment by sundermann** [comment]
Setup git clang format locally which should format it for you. Squash all the fixup commits into one, rebase and apply `clang-format`

**Comment by jeannich** [comment]
> Setup git clang format locally which should format it for you. Squash all the fixup commits into one, rebase and apply `clang-format`

OK thanks, I'll try that

**Comment by jeannich** [comment]
Hi, the voter is not happy with my merge commit:
ERROR: This PR contains 1 merge commit(s)

Can someone assist to fix this? I have no idea how to "remove" this merge commit.

**Comment by garbear** [comment]
Your PR branch is at the commit: [639c3fae2566a17892f9cfb33cb915001513040b](https://github.com/xbmc/xbmc/commit/639c3fae2566a17892f9cfb33cb915001513040b)

You're based against an old branch (v21.0-rc2), so you'll want to get up-to-date. The best way is hard-reset to where you want your base (usually `upstream/master`), then re-apply your work against the fresh branch.

**Comment by fuzzard** [comment]
You need to rebase this onto master. I've changed the targetted branch of the PR. PRs must be submitted against master before any backport is an option

**Comment by jeannich** [comment]
> You need to rebase this onto master. I've changed the targetted branch of the PR. PRs must be submitted against master before any backport is an option

Wow, destination branch changed !
OK, I need to figure out how to handle this ... (no idea for now)

**Comment by fuzzard** [comment]
If need be, you can just close and reopen a new PR targetting master based on master if that's easiest for you.

Otherwise, just rebase this branch on master, and force  push back to this branch is sufficient

**Comment by jeannich** [comment]
Seems impossible to reconcile with branch master, so I close this PR.
Here is the new one: [26067](https://github.com/xbmc/xbmc/pull/26067)

**Pull request closed without merging**